src/autowiring/test/AutoSignalTest.cpp

```diff
@@ -122,3 +122,121 @@ TEST_F(AutoSignalTest, RaiseASignalWithinASlotTest) {
   ASSERT_TRUE(handler_called1) << "Handler 1 was not called on a stack-allocated signal";
   ASSERT_TRUE(handler_called2) << "Handler 2 was not called on a stack-allocated signal";
 }
+
+TEST_F(AutoSignalTest, NodeRemoval) {
+  autowiring::signal<void(void)> signal1;
+  autowiring::signal<void(void)> signal2;
+
+  bool handler_called1 = false;
+  bool handler_called2 = false;
+
+  auto* registration1 = signal1 += [&] { handler_called1 = true; };
+  auto* registration2 = signal2 += [&] { handler_called2 = true; };
+  
+  ASSERT_ANY_THROW(signal1 -= registration2) << "Removing a registration from a different signal than it was registered to failed to throw an exception";
+
+  delete registration1;
+  signal1();
+  signal2();
+  
+  ASSERT_FALSE(handler_called1) << "Handler1 was called after being unregistered";
+  ASSERT_TRUE(handler_called2) << "Handler2 was removed after an invalid -= operation";
+}
+
+TEST_F(AutoSignalTest, CallOrdering) {
+  autowiring::signal<void(void)> signal1;
+
+  bool handler_called1 = false;
+  bool handler_called2 = false;
+
+  //handlers are inserted at the begining, so this will be called last.
+  signal1 += [&] {
+    ASSERT_TRUE(handler_called2);
+    handler_called1 = true;
+  };
+  signal1 += [&] { handler_called2 = true; };
+
+  signal1();
+
+  ASSERT_TRUE(handler_called1) << "Handler1 was called after being unregistered";
+  ASSERT_TRUE(handler_called2) << "Handler2 was removed after an invalid -= operation";
+}
+
+TEST_F(AutoSignalTest, CallInsertion) {
+  autowiring::signal<void(void)> signal1;
+
+  bool handler_called1 = false;
+  bool handler_called2 = false;
+
+  auto* registration1 = signal1 += [&](void) { handler_called1 = true; };
+
+  //when += to a registration object, the new one is appended.
+  *registration1 += [&](void) {
+    ASSERT_TRUE(handler_called1);
+    handler_called2 = true; };
+
+  signal1();
+
+  ASSERT_TRUE(handler_called1) << "Handler1 was called after being unregistered";
+  ASSERT_TRUE(handler_called2) << "Handler2 was removed after an invalid -= operation";
+}
+
+TEST_F(AutoSignalTest, SelfReferencingCall) {
+  typedef autowiring::signal<void(int)> signal_t;
+  signal_t signal1;
+
+  bool handler_called1 = false;
+  int magic_number = 123;
+
+  //The main test is just if this thing will compile
+  signal_t::registration_t* registration1 =
+  signal1 += [&](autowiring::internal::signal_node_base* reg, int magic) {
+    ASSERT_EQ(magic, magic_number);
+    ASSERT_EQ(registration1, reg);
+    handler_called1 = true;
+  };
```

Comment:
Tab this part out so people don't miss the fact that lines 192 to 197 are all the same statement.